Make it possible to abort the pull-stream when there are no listeners left

Happy0 opened the pull request:
Sorry about all the pull requests I've been throwing your way recently @mmckegg =p. Look at them whenever - no expectations / rush / pressure!

When I was doing some work to remove listeners to observables in `ssb-chess` recently, I noticed that the streams for the mutant-pull-reduce  lists underlying my chess games weren't getting closed (I'd added a print statement to the start of the stream, and was wondering why it was still printing when the observable had been unlistened.) I ended up with a bunch of unclosed streams for the same source (the live feed with timestamp gt than something.)

I'm not sure if the messages were still getting printed, regardless of the 'pauser', because it was from a live stream and they were getting pushed anyway.

I'd like to remove the streams for games entirely when the page for the boards are navigated away from, and open a new observable when the game (or list of mini-boards for games) is returned to.

Does adding an option to abort when there are no listeners left, rather than pause, make sense to you?

mmckegg commented:
@Happy0

The problem is we need to handle the case where a reducer is unbound and then re-bound again later (e.g. an item is removed from the DOM, but then re-added later)

But this is complicated because if the item is re-added, we will want to catch up on the events that have happened since it was removed. So the pause stream seemed like an easy answer, but yes you are right that it means you end up with dangling streams (particularly in realtime situations).

I think the pauser should be replaced with an aborter (as you have proposed), but we need a mechanism for resuming the stream somehow if a reducer is re-bound again later.

It could be that the reducer hangs on to a reference of the last item that it read (before abort), and then on resume, will pass that item to the stream function. It's hard to get this right 😢 

---

One idea would be that rather than passing the reducer an stream (like now), you pass it a function to get a stream (`getStream`). The reducer would only call this function once it was bound/listened to. It would abort this stream when unlistened, and then call this function again if the item was re-bound/listened again.

It could pass the last item that it received to the `getStream` function so that the stream could resume from a given point if needed.

```
var reducer = MutantPullReduce(getStream, (result, item) => {
  // do something!
  return result
}, {
  startValue: []
})

function getStream (lastItem) {
  if (lastItem) {
    // resume (with gt/lt option)
    // or if return nothing, then reducer doesn't resume
  } else {
    // stream from start
  }
}
```

This change would obviously be a major breaking! But I think we need to do something to improve this situation.

Happy0 commented:
Sounds good to me :)

Should the `getStream` function accept a parameter with the last message so
that it can use that last message to continue the stream from the last
position in a generalised way?

In scuttlebutt we have gte and the sequence number, but other things using
this library with pull streams might use some other mechanism.

On 5 Jan 2018 22:53, "Matt McKegg" <notifications@github.com> wrote:

> @Happy0 <https://github.com/happy0>
>
> The problem is we need to handle the case where a reducer is unbound and
> then re-bound again later (e.g. an item is removed from the DOM, but then
> re-added later)
>
> But this is complicated because if the item is re-added, we will want to
> catch up on the events that have happened since it was removed. So the
> pause stream seemed like an easy answer, but yes you are right that it
> means you end up with dangling streams (particularly in realtime
> situations).
>
> I think the pauser should be replaced with an aborter (as you have
> proposed), but we need a mechanism for resuming the stream somehow if a
> reducer is re-bound again later.
>
> It could be that the reducer hangs on to a reference of the last item that
> it read (before abort), and then on resume, will pass that item to the
> stream function. It's hard to get this right 😢
> ------------------------------
>
> One idea would be that rather than passing the reducer an stream (like
> now), you pass it a function to get a stream (getStream). The reducer
> would only call this function once it was bound/listened to. It would abort
> this stream when unlistened, and then call this function again if the item
> was re-bound/listened again.
>
> It could pass the last item that it received to the getStream function so
> that the stream could resume from a given point if needed.
>
> var reducer = MutantPullReduce(getStream, (result, item) => {
>   // do something!
>   return result
> }, {
>   startValue: []
> })
>
> function getStream (lastItem) {
>   if (lastItem) {
>     // resume (with gt/lt option)
>     // or if return nothing, then reducer doesn't resume
>   } else {
>     // stream from start
>   }
> }
>
> This change would obviously be a major breaking! But I think we need to do
> something to improve this situation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mmckegg/mutant-pull-reduce/pull/1#issuecomment-355687414>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABF5XBfvLTuG2wOvivsDARWdGeRc5OTqks5tHqfLgaJpZM4RU9zp>
> .
>

Happy0 commented:
Oh wait, that's what you said :p

I'll update this PR to do that :)

Happy0 committed:
BREAKING: A new strategy to handle the observable being re-subscribed
to after all subscribers have previously unsubscribed.

With the previous strategy, if a live stream was unsubscribed from
and then re-subscribed to, the subscriber would not have been able to
see the values it has missed inbetween.

The new approach allows the caller to specify how to resume the stream
from the appropriate point by taking a function that accepts the latest
value and return a new stream. See the README.md document for more
details.

Happy0 commented:
@mmckegg I update this PR with the approach we discussed.

I want to write some unit tests for it and add them to the PR too, but I just wanted to push to make sure the general approach looks okay?

Happy0 committed:
Remove pull-pause dependency.

Happy0 committed:
Start adding unit tests.

Happy0 commented:
any idea why 'unsubscribe' is 'null' here? https://github.com/mmckegg/mutant-pull-reduce/pull/1/files#diff-47b1537e869352957a20a39df5226cddR74

I want to remove the listener once a certain point in the stream is reached (and i've unsubscribed from within a handler for a mutant observable before.)

Is there some weird scoping thing going on because of browserfy? =<

Happy0 commented:
Oh.. that's what nextTick is for :o (sorry for being spammy0 =p)

Happy0 committed:
nextTick

Happy0 commented:
Ha. This is actually tougher to test than I thought - I guess because the stream happens asynchronously and the 'abort' is (probably) being called after the stream has fully drained.

I'll check out the unit tests in pull-abortable and see how it deals with the async nature of pull-streams when testing :P

Happy0 committed:
Progress

Happy0 committed:
making progress.

Happy0 commented:
Okay, now I'm stuck and it's becoming no fun :P.

Going to drop it for a bit and come back to it in a day or two =p.

mmckegg commented on index.js:
```diff
@@ -31,30 +48,30 @@ module.exports = function (stream, reducer, opts) {
+   */
+  function startStream() {
+    pull(
+      getStream(binder.getValue()),
```
I think this value should be the last item that came through the stream, rather than the reduced state.

Happy0 replied:
True =]. It occurred to me when I hacked the index.js file to bits to try and work out why the unit tests aren't passing.

I've updated the PR to use the last stream value. The unit tests are still failing - will try and work out what the deal is with those later =p.

Happy0 committed:
Track the last stream value separately.
